### PR TITLE
Added AllMatch, AnyMatch, NoneMatch generic range matchers

### PR DESCRIPTION
**Pull request by urielgarciarivas**
## Description
Added AnyMatch, AllMatch and NoneMatch generic range matchers which take single matcher during construction, and their `match` method take a generic range, using the new generic matcher support.

## GitHub Issues
Addressing issue #2109 .

### CODE REVIEW

File: src/catch2/matchers/catch_matchers_all.hpp
Comment (horenmar) [inline src/catch2/matchers/catch_matchers_all.hpp]: This is not a good filename given the contents. "Logical" would be better, but it still doesn't feel ideal.

File: src/catch2/matchers/catch_matchers_generic.hpp
Comment (horenmar) [inline src/catch2/matchers/catch_matchers_generic.hpp]: There is no overload set, so this does not need to be conditionally enabled.

I don't think there is even a need to static assert on this -- if it quacks like a matcher, there is no reason not to use it.

Comment (horenmar) [reply src/catch2/matchers/catch_matchers_generic.hpp]: (This obviously goes for the other functions as well)

File: src/catch2/matchers/catch_matchers_generic.hpp
Comment (horenmar) [inline src/catch2/matchers/catch_matchers_generic.hpp]: I don't think this header uses either of these two.

File: src/catch2/matchers/catch_matchers_generic.hpp
Comment (horenmar) [inline src/catch2/matchers/catch_matchers_generic.hpp]: You shouldn't need these, range-for already does the requisite ADL lookup

**Review by horenmar — CHANGES_REQUESTED** [review]
Thanks.

The PR is off to a good start, but some changes are required and there need to be tests.

File: src/catch2/matchers/catch_matchers_all.hpp
Comment (urielgarciarivas) [reply src/catch2/matchers/catch_matchers_all.hpp]: Do you mean I need to create a file for each matcher? e. g. `catch_matchers_addall.hpp`

File: src/catch2/matchers/catch_matchers_generic.hpp
Comment (urielgarciarivas) [reply src/catch2/matchers/catch_matchers_generic.hpp]: You are right!! Haha thanks

File: src/catch2/matchers/catch_matchers_all.hpp
Comment (horenmar) [reply src/catch2/matchers/catch_matchers_all.hpp]: Not necessarily. But I don't think people would expect to find `AllMatch`, `AnyMatch`, `NoneMatch` in file named "matchers generic", when the documentation already talks about "Generic Matchers" as a way of defining new matchers. Ideally there would be a good name for them as a group, but I can't think of any.

Placing them each into a separate file works too I guess, e.g. `catch_matchers_allmatch.hpp`

Comment (urielgarciarivas) [reply src/catch2/matchers/catch_matchers_all.hpp]: What about quantifiers? `catch_matchers_quantifiers.hpp` which includes `AllMatch`, `AnyMatch` and `NoneMatch`.

Comment (horenmar) [reply src/catch2/matchers/catch_matchers_all.hpp]: I won't say that I think it is great, but since I've already gotten "quantifiers" as a suggestion from someone else, it is apparently not as unintuitive as I though.

So, sure, go for it. 👍

Comment (urielgarciarivas) [reply src/catch2/matchers/catch_matchers_all.hpp]: Great, I'll update it!

**Comment by urielgarciarivas** [comment]
I need to create tests in `test/SelfTest/UsageTests/MatchersRanges.tests.cpp` or should I create them somewhere else?

**Review by horenmar — APPROVED** [review]

**Comment by horenmar** [comment]
Put them into `MatchersRanges.tests.cpp` . I intend to refactor (mostly split them apart) the matchers tests, but there is no reason for you to worry about that.

**Comment by horenmar** [comment]
You need to update [Approval tests](https://github.com/catchorg/Catch2/blob/devel/docs/contributing.md#testing-your-changes)


Also you should test that the matchers support ranges that require ADL, and ranges that use different type for `begin` and `end`.

**Comment by horenmar** [comment]
Okay, so what's missing now:
* Document the changes
* Tests
  1) Can the matcher handle a type that requires its specific ADL begin found?
  2) Can the matcher handle a type that returns different types for `begin` and `end`?
  3) Does the matcher shortcircuit?

I can take it from here if you want.

**Comment by urielgarciarivas** [comment]
What would happen to my PR? Would you merge this PR or should this PR wait until you add what's missing?

**Comment by horenmar** [comment]
I would add the tests on top of your branch and then merge it when it is done.

**Comment by urielgarciarivas** [comment]
Please do! Thank you very much!

**Comment by horenmar** [comment]
Okay, tests should be there.

I'll wait for the CI to finish, and then rebase + cleanup the MR once again before merging.

**Comment by urielgarciarivas** [comment]
Closes #2109

**Pull request merged**
